### PR TITLE
Change the codebase to use the new attributes table

### PR DESCRIPTION
**Pull request by orange-vertex**
New methods for attribute table:
    clear()
    getSelAvg() could potentially store that somewhere?
    perhaps maintain a key index?
    created a dummy column for the key reference
Selected set and some other containers now store keys not indices
The local (pointmap/shapemap) field m_display_params is never updated so it has been removed

Problematic tests:
In cases where the columns have been added to the attribute table in the wrong order it is impossible to reconcile with the base binary without keeping track of the insertion order:
    VGA Angular has wrong order of placing columns ("Angular Node Count" was placed after "Angular Total Depth") [regression case "vga_angular"]
    Data to Axial Conversion has wrong order where the "Data Map Ref" column is added first and its index kept but then superseeded (by the many Choice columns) [regression case: "convert_axial_data_axial_copy_attributes"]
    Segment to Axial Conversion has wrong order where the "Data Map Ref" column is added first and its index kept but then superseeded (by the Axial Line Ref column).[ "convert_segment_data_segment_copy_attributes"]

In the original implementaion the maximum of the column is set at 0 and the total may go under -1. In the new one the maximum is set to -1 and the total is reset if it is found to be under 0. In the case of a new column where all the values are -1, if it's never updated then the minumm value for both is -1, the max 0 in the original and -1 in the new, and the total -1*(number of rows) and -1:
    This happens in case "convert_drawing_segment" thus, a new case is provided called "convert_drawing_segment_only_map"

PointMaps are not copyable anymore as they contain unique_ptrs to the attr table and thus the relevant unit tests from testpointmap have been removed

It appears that with the original implementation, the minimum value of a column has no way to be increased (apart from re-iterating the column) if found once to be very small. Case in point is the agent analysis which increases the value of pixels by 1. As that is the lowest value ever sent to the column it remains its m_min even if the cell with that value is increased again. This is what we see at regression test case "agents_line_of_sight_length_look" where the lowest "Gate count" value is actually 4 but m_min in the binary file is 1. The new implementation does not suffer from this and thus when we compare only the attribute values (new regression test "agents_line_of_sight_length_look_only_map") the two are identical.

The unit tests "ShapeMap::getAllLinesWithColour" and "ShapeMap::getAllPolygonsWithColour" in testshapemaps have been modified to reflect the proper values of the colours of the ref column. As the test is rudimentary, in the original implementation the actual "normalised" value of the Ref column was never calculated as the process was never triggered (say by a "setDisplayColumn" call). Thus the shapes in the unit test were all given the same default gray 0x007f7f7f. This is fixed in the new implementation

### CODE REVIEW

File: depthmapX/views/3dview/3dview.cpp [hunk 1]
```diff
@@ -459,9 +460,9 @@ void Q3DView::ReloadPointData()
       //
       m_pixels.clear();
       PointMap& map = pDoc->m_meta_graph->getDisplayedPointMap();
-      AttributeTable& table = map.getAttributeTable();
-      for (int i = 0; i < table.getRowCount(); i++) {
-         PixelRef pix = table.getRowKey(i);
+      dXreimpl::AttributeTable& table = map.getAttributeTable();
+      for (auto iter = table.begin(); iter != table.end(); iter++) {
```
Comment (blsemo) [inline depthmapX/views/3dview/3dview.cpp]: For correctness, shouldn't that be `for (const auto& iter = ...`?

File: depthmapX/views/plotview/plotview.cpp [hunk 2]
```diff
@@ -269,15 +272,36 @@ bool QPlotView::Output(QPainter *pDC, QGraphDoc *pDoc, bool screendraw)
...
+   LayerManagerImpl& layers = pDoc->m_meta_graph->getLayers();
    int displaycol = pDoc->m_meta_graph->getDisplayedAttribute();
 
    QRect rect = QRect(0, 0, width(), height());
    int mindim = __min(rect.width(),rect.height());
-   int viscount = table.getVisibleRowCount();
-   if(!viscount) m_x_axis = m_y_axis = 0;
 
-   int spacer = floor(2.0 * sqrt(double(mindim)/double(viscount != 0 ? viscount : 1)));
+   // do all the calculations here for the moment
```
Comment (blsemo) [inline depthmapX/views/plotview/plotview.cpp]: Hmm, that looks like something that should be in a helper class/method

File: depthmapX/views/plotview/plotview.cpp [hunk 3]
```diff
@@ -710,29 +735,29 @@ void QPlotView::mouseReleaseEvent(QMouseEvent *e)
    }
    m_selecting = false;
 
-   size_t xfloor = idx_x.searchfloorindex(ValuePair(-1,dataX(m_drag_rect_a.left()-2)));
-   size_t xceil = idx_x.searchceilindex(ValuePair(-1,dataX(m_drag_rect_a.right()+2)));
-   size_t yfloor = idx_y.searchfloorindex(ValuePair(-1,dataY(m_drag_rect_a.bottom()+2)));
-   size_t yceil = idx_y.searchceilindex(ValuePair(-1,dataY(m_drag_rect_a.top()-2)));
+   dXreimpl::AttributeTable& table = pDoc->m_meta_graph->getAttributeTable();
```
Comment (blsemo) [inline depthmapX/views/plotview/plotview.cpp]: We seem to have to do this more than once - so maybe AttributeTable should have a method to make that cleaner and easier.

File: salalib/attributetable.cpp [hunk 4]
```diff
@@ -413,9 +433,20 @@ namespace  dXreimpl {
         layerManager.write(stream);
         int colCount = (int)m_columns.size();
         stream.write((char *)&colCount, sizeof(int));
-        for (size_t i = 0; i < m_columns.size(); ++i)
-        {
-            m_columns[i].write(stream, (int)i);
+
+        // TODO: For binary compatibility write the columns in alphabetical order
```
Comment (blsemo) [inline salalib/attributetable.cpp]: This still makes me angry!

File: salalib/attributetable.cpp [hunk 5]
```diff
@@ -441,24 +478,54 @@ namespace  dXreimpl {
 
     }
 
+    // TODO: Compatibility. Very inefficient method to retreive a column's index
```
Comment (blsemo) [inline salalib/attributetable.cpp]: Couldn't we just get the name for the index, and then count the steps to that name in the columns mapping?

File: salalib/salaprogram.cpp [hunk 6]
```diff
@@ -1562,14 +1558,14 @@ SalaObj SalaCommand::evaluate(int& pointer, SalaObj* &p_obj)
                case SalaObj::S_FMARK:
                   {
                      param.ensureNone();
-                     AttributeTable *table = obj.getTable();
-                     data = m_program->marks[(obj.type == SalaObj::S_POINTMAPOBJ) ? table->getRowid(obj.data.graph.node) : obj.data.graph.node];
+                     dXreimpl::AttributeTable *table = obj.getTable();
```
Comment (blsemo) [inline salalib/salaprogram.cpp]: `table` is not used here, is it?

File: salalib/salaprogram.cpp [hunk 7]
```diff
@@ -1656,7 +1652,9 @@ SalaObj SalaCommand::connections(SalaObj graphobj, SalaObj param)
       }
    }
    else {
-      const Connector& connector = graphobj.data.graph.map.shape->getConnections()[graphobj.data.graph.node];
+      int idx = std::distance(graphobj.data.graph.map.shape->getAllShapes().begin(),
```
Comment (blsemo) [inline salalib/salaprogram.cpp]: I can't see the context here - do we need the `std::distance` here or are we in a case where keys are just indices?

File: salalib/segmmodules/segmangular.cpp [hunk 8]
```diff
@@ -132,21 +132,23 @@ bool SegmentAngular::run(Communicator *comm, const Options &options, ShapeGraph
                 anglebins.erase(iter);
             }
         }
+        dXreimpl::AttributeRow &row =
+            attributes.getRow(dXreimpl::AttributeKey(depthmapX::getMapAtIndex(map.getAllShapes(), i)->first));
```
Comment (blsemo) [inline salalib/segmmodules/segmangular.cpp]: I think it's safe to assume here that index is key...

File: depthmapX/views/plotview/plotview.cpp (hunk 2 above)
Comment (orange-vertex) [reply depthmapX/views/plotview/plotview.cpp]: It should be indeed, but I tried looking into this and the proper solution, where these stats would be stored in the attribute table like the other stats, and it's quite complex to handle in this PR, due to having to update them every time the layers change. Thus, I have just updated the comment to fix at a later stage.

**Comment by orange-vertex** [comment]
> One thing that needs to be done in detail, I am afraid is going through the code and find the places where the keys are really just zero based indices (I think that is everything that doesn't get used in VGA?) and remove all the `findIndexInMap` calls where they are not necessary, as they are massively expensive.

I did go over these and reduced the map search in as many as I could find. The problem is that these indices are used in a lot of places which would need to be changed to keys so as to fully solve the problem. The main pain point is that the connections of axial and segment lines are index-based so anything that touches that requires dipping into the indices and will likely make everything slower. There are other such cases (with indices) such as the functions to find points in polygons, various local intermediate containers and other bits that would need to be cleaned up. I would pin these to a different development branch, an effort to make depthmap's various modelling (isovist creation) and analysis (integration/depths) functions faster. In such a branch, changing the connectors should be the primary target.

**Pull request merged**